src/mixins/yield-erc/ERC20Vault.sol

```diff
@@ -0,0 +1,149 @@
+
+        baseUnit = 10**decimals;
+    }
+
+    /*///////////////////////////////////////////////////////////////
+                        DEPOSIT/WITHDRAWAL LOGIC
+    //////////////////////////////////////////////////////////////*/
+
+    /// @notice Emitted after a successful deposit.
+    /// @param user The address that deposited into the Vault.
+    /// @param underlyingAmount The amount of underlying tokens that were deposited.
+    event Deposit(address indexed user, uint256 underlyingAmount);
+
+    /// @notice Emitted after a successful withdrawal.
+    /// @param owner The address that withdrew from the Vault.
+    /// @param to The destination for withdrawn tokens.
+    /// @param underlyingAmount The amount of underlying tokens that were withdrawn.
+    event Withdraw(address indexed owner, address indexed to, uint256 underlyingAmount);
+
+    /// @notice Deposit a specific amount of underlying tokens.
+    /// @param to The address to receive shares corresponding to the deposit
+    /// @param underlyingAmount The amount of the underlying token to deposit.
+    function deposit(address to, uint256 underlyingAmount) external virtual returns (uint256 shares) {
+        shares = underlyingAmount.fdiv(exchangeRate(), baseUnit);
+        // Determine the equivalent amount of shares and mint them.
+        _mint(to, shares);
+
+        emit Deposit(to, underlyingAmount);
+
+        // Transfer in underlying tokens from the user.
+        // This will revert if the user does not have the amount specified.
+        underlying.safeTransferFrom(msg.sender, address(this), underlyingAmount);
+
+        afterDeposit(underlyingAmount);
+    }
+
+    /// @notice Withdraw a specific amount of underlying tokens.
+    /// @param to The address to receive underlying tokens corresponding to the withdrawal.
+    /// @param underlyingAmount The amount of underlying tokens to withdraw.
+    function withdraw(address to, uint256 underlyingAmount) external virtual returns (uint256 shares) {
+        shares = underlyingAmount.fdiv(exchangeRate(), baseUnit);
+
+        // Determine the equivalent amount of shares and burn them.
+        // This will revert if the user does not have enough shares.
+        _burn(msg.sender, shares);
+
+        emit Withdraw(msg.sender, to, underlyingAmount);
+
+        // Withdraw from strategies if needed and transfer.
+        beforeWithdraw(underlyingAmount);
+
+        underlying.safeTransfer(to, underlyingAmount);
+    }
+
+    /// @notice Redeem a specific amount of shares for underlying tokens.
+    /// @param to The address to receive underlying tokens corresponding to the withdrawal.
+    /// @param shareAmount The amount of shares to redeem for underlying tokens.
+    function redeem(address to, uint256 shareAmount) external virtual returns (uint256 underlyingAmount) {
+
+        // Determine the equivalent amount of underlying tokens.
+        underlyingAmount = shareAmount.fmul(exchangeRate(), baseUnit);
```

Comment:
In `deposit` and `withdraw` you define `shares = underlying /↓ exchangeRate`, so here you need to use `underlying = shares *↑ exchangeRate`.

Reply:
yup, ABDK also pointed this out in the vaults audit

Reply:
@transmissions11 can you patch in your fix for vaults?

Reply:
i need to impl a fmulUp in solmate first, will do in a bit

Reply:
I would reverse the exchange rate, use fmul for deposit and withdraw, and fdivup for the rest. I've found it makes for an easier flow to avoid fdiv for as long as possible, and having a consistent approach helps integrations. Personal preference, though.